daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ElementUnparser.scala

```diff
@@ -417,39 +418,56 @@ sealed trait RegularElementUnparserStartEndStrategy
             UnparseError(Nope, One(state.currentLocation), "Expected element start event for %s, but received %s.",
               erd.namedQName.toExtendedSyntax, event)
           }
-          val res = event.info.element
-          val mCurNode = state.currentInfosetNodeMaybe
-          if (mCurNode.isDefined) {
-            val c = mCurNode.get.asComplex
-            Assert.invariant(!c.isFinal)
-            if (c.maybeIsNilled == MaybeBoolean.True) {
-              // cannot add content to a nilled complex element
-              UnparseError(One(erd.schemaFileLocation), Nope, "Nilled complex element %s has content from %s",
-                c.erd.namedQName.toExtendedSyntax,
-                res.erd.namedQName.toExtendedSyntax)
-            }
-            c.addChild(res, state.tunable)
-          } else {
-            val doc = state.documentElement
-            doc.addChild(res, state.tunable) // DIDocument, which is never a current node, must have the child added
-            doc.isFinal = true // that's the only child.
-          }
-          res
+          event.info.element
         } else {
           Assert.invariant(state.withinHiddenNest)
           // Since we never get events for elements in hidden contexts, their infoset elements
           // will have never been created. This means we need to manually create them
-          val e = if (erd.isComplexType) new DIComplex(erd) else new DISimple(erd)
-          e.setHidden()
-          state.currentInfosetNode.asComplex.addChild(e, state.tunable)
-          e
+          val hiddenElem = if (erd.isComplexType) new DIComplex(erd) else new DISimple(erd)
+          hiddenElem.setHidden()
+          hiddenElem
+        }
+
+      // now add this new elem to the infoset
+      val parentNodeMaybe = state.currentInfosetNodeMaybe
+      if (parentNodeMaybe.isDefined) {
+        val parentComplex = parentNodeMaybe.get.asComplex
+        Assert.invariant(!parentComplex.isFinal)
+        if (parentComplex.maybeIsNilled == MaybeBoolean.True) {
+          // cannot add content to a nilled complex element
+          UnparseError(One(erd.schemaFileLocation), Nope, "Nilled complex element %s has content from %s",
+            parentComplex.erd.namedQName.toExtendedSyntax,
+            newElem.erd.namedQName.toExtendedSyntax)
         }
 
+        // We are about to add a child to this complex element. Before we do
+        // that, if the last child added to this complex is a DIArray, and this
+        // new child isn't part of that array, that implies that the DIArray
+        // will have no more children added and should be marked as final, and
```

Comment:
> So there's some confusion here around the role of isFinal in my mind.

They do feel separate, but are actually closely related. isFinal is only set for elements when unparseEnd is called, which means the element must have been fully unparsed and can be removed. The place where this sort feels question is with suspensions. The unparse process looks like this:

1. parseBegin - create an Element and add to infoset
2. Unparse body, potentially creating a Suspension
3. parseEnd - mark element as isFInal and free

So I think a concern is that we might unparse the body and create a suspension, but then immediately call parseEnd and free the element. However, creating a suspension clones the current infoset, so while the element is freed from the "main" infoset, the suspension still has a pointer to it and can do the normal processing.

> If we wait for an entire array to be isFinal before we free any of its child elements, then we cannot truly stream a large array. Freeing element with index N-1 upon unparse of element N seems essential to obtain streaming behavior.

We do not wait for an array to be final before freeing any of it's children. We free the children as soon as they are done unparsing, because those children will call unparseEnd, get marked as final, and they will be freed. Only when all those children are finished and we add something after the array is the arrary marked as isFInal and the array can be freed.

> I would expect we would pass the array index down to the freeChildIfNoLongerNeeded call, since without that it can't decide to release things based on index.

We do not currently free anything based on index. We free things as soon as possible, including children of arrays.

The one case were don't free things (in both parse and unparse) is when something is used in an expression. So large arrays are only a problem if they are use in expressions. And even then, we still only keep the things in those arrays that are used in expressions. So it's still wasteful, and large arrays can cause problems, but it's certainly less of an issue.

There is a proposal to add a new dfdlx property that gives hints to Daffodil to say that if an array with this property is used in an expression (and thus we can't free it's contents) then we only need to keep the most recent X elements. This isn't implemented yet. Perhaps this can be postponed for a future release?

Reply:
So, if I have a format which is a giant array, e.g., PCAP format, the elements of the array are all referenced by expressions because the header of each packet needs the length of the packet contents. So we'll never free anything (pretty much. We might be able to discard a few header elements in each packet, but we won't be able to discard the packet payload contents) when parsing or unparsing PCAP? If that's the case, then this isn't really solving people's streaming issue until this dfdlx extension property telling it to prune arrays is implemented? 
Or.... since the suspension to fill in the length of the packet in the packet header is evaluated, at that point do we mark everything final and so that we now know we can release it?

Reply:
I'd have to do some testing, but I think this would actualy stream somewhat okay. The PCAP schema essentially looks like this:
```xml
<PCAP>
  <GlobalHeader>...<GlobalHeader>
  <Packet>...</Packet>
  ...
  <Packet>...</Packet>
</PCAP>
```
Named down steps are the only thing that cause elements to be marked as "referenced by an expression" and thus not freeable. And the PCAP schema, nothing actually references "Packet" in a down step. Lots of children of the Packet element do get referenced, but not the Packet itself. So when a Packet element is marked as isFinal, we will immediately free it. So I suspect we'll only ever have one Packet (and all of it's children) in memory at a time. Which should be very reasonble.

This does exlucding things like suspensions which might hold on to packets, but hopefully the earlier suspension evaluation will mitigate that.

This also doesn't handle the fact the there there is going to be a DIArray with size N where N is the number of packets. Most of the elements in that array will be null because they'll be freed, but it will still take up a good chunk of memory for large arrays. Note that a schema could minimize this a bit by "chunking" the array. E.g.
```xml
<xs:element name="PacketChunk" maxOccurs="unbounded">
  <xs:complexType>
    <xs:sequence>
      <xs:element name="Packet" maxOccurs="10000" ... />
    </xs:sequence>
  </xs:complexType>
</xs:element>
```
It's a bit ugly, but is a potential work around for very large arrays without having to compact them.

Note that we could add a feature to compact them after they grow to some size and have enough elements freed, but there are things like fn:count() and index offsets that need to be considered, so if we do want to do that, it's probably best left for another change.

I'm also not argueing that the extension to hint to clean up arrays isn't necessary. But I do think that this change does resolve some streaming issues without requiring additional extension properties be added.

Reply:
Ok. Thanks. This is what I didn't understand. I do understand the array chunking trick, and how that would perhaps work to minimize the array. But the real problem is not a large spine of an array, but the contents of the array, at least for now.